### PR TITLE
[NPU] add Ascend NPU support for higgs-audio-v3-tts-4b

### PR DESCRIPTION
**Pull request by chx96642264**
## Motivation

`higgs-audio-v3-tts-4b` was only testable on CUDA: the pipeline config pinned
`device="cuda"` for all three stages, the batched sampler imported the fused
`sgl_kernel` top-k/top-p renorm kernels unconditionally (an import that pulls
in CUDA-only code and fails on Ascend NPU), and the codec acoustic encoder /
vocoder decode were wrapped with `torch.compile` and relied on CUDA graph
capture — none of which exist on Ascend NPU (Atlas 910B).

So some adaptation work was done on the NPU in this pr.

## Modifications

- `sampler.py`: platform-aware top-k/top-p renorm kernels and platform-resolved device defaults.
- `config.py` / `stages.py`: use `current_platform.device_type`
  instead of hardcoded `"cuda"`; gate `torch.compile` and codec decode CUDA
  graphs on platform support (eager on NPU).
- `engine_builder.py`: `sampling_backend="pytorch"` (can work on every backend).

## Related Issues

N/A

## Accuracy Test

Test in sglang v0.5.18.

Evaluated on `zhaochenyang20/seed-tts-eval-arrow` (SeedTTS EN, 1088 samples,
voice-clone, non-streaming), with Qwen3-ASR-1.7B as the transcription oracle,
on Ascend NPU (Atlas 910B).

| Metric(concurrency=16) | NPU (this PR) | H200 |
|---|---|---|
| wer_corpus | 0.95% | 1.10% |
| wer_per_sample_mean | 0.91% | 1.08% |
| wer_per_sample_median | 0.00% | 0.00% |
| wer_per_sample_std | 3.4% | 4.1% |

The non-outlier corpus WER (0.95%) is at or below the CUDA reference,
confirming no systematic accuracy regression from the NPU adaptation.

Verification screenshot is as follows:

higgs_tts_wer_new

## Benchmark & Profiling
Here are the performance-related test results. I have also added some test cases for the NPU scenario. 

| Concurrency | Throughput (req/s) | Mean Latency | RTF Mean | audio_s/s |
|---|---|---|---|---|
| 1 | 0.691 | 1.447 s | 0.3424 | 2.948 |
| 2 | 1.394 | 1.433 s | 0.3366 | 6.013 |
| 4 | 2.422 | 1.65 s | 0.3914 | 10.323 |
| 8 | 4.09 | 1.951 s | 0.4638 | 17.407 |
| 16 | 6.269 | 2.536 s | 0.6124 | 26.176 |

Performance verification screenshot is as follows:

higgs_tts_perf_c1
higgs_tts_perf_c2
higgs_tts_perf_c4
higgs_tts_perf_c8
higgs_tts_perf_c16

Test cases verification screenshot is as follows:
higgs_tts_test_new

## Checklist

- [ ] Format your code according with pre-commit.
- [ ] Add unit tests.
- [ ] Update documentation / docstrings / example tutorials as needed.
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the
`run-ci` label. Once labeled, every subsequent push re-triggers CI as
long as the label remains. Use `/tag-and-rerun-ci higgs` or
`/tag-and-rerun-ci moss` or `/tag-and-rerun-ci qwen3-tts` to select a TTS CI
model, and
`/tag-and-rerun-ci fun-asr`, `/tag-and-rerun-ci qwen3-asr` or
`/tag-and-rerun-ci whisper-asr` to select an ASR CI model. One selector from
each family can be combined, for example `/tag-and-rerun-ci moss fun-asr`.
Draft PRs are skipped even if labeled.

### CODE REVIEW

**Comment by ping1jing2** [comment]
i convert this PR to draft due to no description here, please click "Ready to review" after it's ready

File: sglang_omni/models/qwen3_asr/engine_builder.py [hunk 1]
```diff
@@ -123,18 +118,26 @@ def pre_infra_setup(self, checkpoint_dir: str) -> None:
...
-            "enable_torch_compile": self.enable_torch_compile,
+            "enable_torch_compile": enable_torch_compile,
             "torch_compile_max_bs": self.torch_compile_max_bs,
             "mem_fraction_static": self.mem_fraction_static,
             "max_prefill_tokens": 4096,
             "chunked_prefill_size": 4096,
             "sampling_backend": "pytorch",
             "dtype": dtype,
-            "cuda_graph_backend_prefill": CudaGraphBackend.BREAKABLE,
+            "cuda_graph_backend_prefill": (
```
Comment (Makcum888e) [inline sglang_omni/models/qwen3_asr/engine_builder.py]: should we add `current_platform.cuda_graph_backend_prefill` here?

Comment (chx96642264) [reply sglang_omni/models/qwen3_asr/engine_builder.py]: Sorry for including other unrelated code in the draft. This is not my modification.

File: sglang_omni/models/higgs_tts/__init__.py [hunk 2]
```diff
@@ -13,18 +13,23 @@
...
+# gated off on Ascend NPU (no torch.compile backend), so the capability is CUDA-only.
+_supports_torch_compile = not current_platform.is_npu()
+
 CAPABILITIES = ModelCapabilities(
     supports_reference_audio=True,
     supports_batch_vocoder=True,
     supports_streaming_vocoder=True,
     supports_cuda_graph=True,
-    supports_torch_compile=True,
+    supports_torch_compile=_supports_torch_compile,
```
Comment (ping1jing2) [inline sglang_omni/models/higgs_tts/__init__.py]: please check and describe the root cause about why we don't support torch.compile

File: sglang_omni/models/higgs_tts/sampler.py [hunk 3]
```diff
@@ -30,6 +29,62 @@
...
+        -1, (k_safe - 1).unsqueeze(-1)
+    ).squeeze(-1)
+    masked = torch.where(probs < kth.unsqueeze(-1), torch.zeros_like(probs), probs)
+    denom = masked.sum(dim=-1, keepdim=True)
+    return masked / denom.clamp_min(torch.finfo(probs.dtype).tiny)
+
+
+def _top_p_renorm_prob_torch(probs: torch.Tensor, p: torch.Tensor) -> torch.Tensor:
+    """torch fallback for ``sgl_kernel.top_p_renorm_prob`` (Ascend NPU).
+
```
Comment (ping1jing2) [inline sglang_omni/models/higgs_tts/sampler.py]: i suggest to move these func into a separate script file named `npu_fallback.py`

File: sglang_omni/models/higgs_tts/sampler.py [hunk 4]
```diff
@@ -265,10 +322,10 @@ def _sample_independent_batched(
             .to(torch.int32)
             .contiguous()
         )
-        probs = _fused_top_k_renorm(probs, tk)
+        probs = _top_k_renorm(probs, tk)
     if top_p is not None:
         tp = top_p.view(B, 1).expand(B, N).reshape(B * N).to(torch.float32).contiguous()
-        probs = _fused_top_p_renorm(probs, tp)
+        probs = _top_p_renorm(probs, tp)
```
Comment (ping1jing2) [inline sglang_omni/models/higgs_tts/sampler.py]: you are not supposed to change this func name

File: sglang_omni/models/higgs_tts/sampler.py [hunk 5]
```diff
@@ -30,6 +29,62 @@
...
+    it falls back to the equivalent torch implementations above.
+    """
+    if current_platform.device_type == "cuda":
+        try:
+            from sgl_kernel import top_k_renorm_prob, top_p_renorm_prob
+
+            return top_k_renorm_prob, top_p_renorm_prob
+        except ImportError:
+            pass
+    return _top_k_renorm_prob_torch, _top_p_renorm_prob_torch
```
Comment (ping1jing2) [inline sglang_omni/models/higgs_tts/sampler.py]:
```suggestion
def _resolve_renorm_kernels():
    """Return ``(top_k_renorm, top_p_renorm)`` callables for this platform.

    CUDA uses the fused ``sgl_kernel`` kernels to avoid a full-vocab
    ``torch.sort`` in decode. Ascend NPU has no fused renorm kernels so
    it falls back to the equivalent torch implementations above.
    """
    if current_platform.device_type == "cuda":
        from sgl_kernel import top_k_renorm_prob, top_p_renorm_prob

        return top_k_renorm_prob, top_p_renorm_prob
    if current_platform.is_npu():
         from .npu_fallback import top_k_renorm_prob, top_p_renorm_prob
         return top_k_renorm_prob, top_p_renorm_prob

    return _top_k_renorm_prob_torch, _top_p_renorm_prob_torch
```

File: sglang_omni/models/higgs_tts/sampler.py [hunk 6]
```diff
@@ -30,6 +29,62 @@
...
+
+
+_top_k_renorm, _top_p_renorm = _resolve_renorm_kernels()
```
Comment (ping1jing2) [inline sglang_omni/models/higgs_tts/sampler.py]:
```suggestion
_fused_top_k_renorm, _fused_top_p_renorm = _resolve_renorm_kernels()
```

**Review by ping1jing2 — APPROVED** [review]

File: sglang_omni/models/higgs_tts/sampler.py (hunk 4 above)
Comment (chx96642264) [reply sglang_omni/models/higgs_tts/sampler.py]: Done. I will keep the original _fused_top_k_renorm / _fused_top_p_renorm names.

File: sglang_omni/models/higgs_tts/__init__.py (hunk 2 above)
Comment (chx96642264) [reply sglang_omni/models/higgs_tts/__init__.py]: Done. NPU does not support torch.compile because the stage_workers.tts_frontend process fails during startup: torch_npu's inductor backend has incomplete support for complex dynamic-shape graphs and crashes during the codegen stage. See the traceback below for details.

higgs_tts_not_compile

File: sglang_omni/models/higgs_tts/sampler.py (hunk 3 above)
Comment (chx96642264) [reply sglang_omni/models/higgs_tts/sampler.py]: Done. I have migrated the fallback implementations to npu_fallback.py as suggested.

File: sglang_omni/models/higgs_tts/sampler.py (hunk 6 above)
Comment (chx96642264) [reply sglang_omni/models/higgs_tts/sampler.py]: Done. Please review the changes to see if they meet your requirements.

File: sglang_omni/models/higgs_tts/sampler.py (hunk 5 above)
Comment (chx96642264) [reply sglang_omni/models/higgs_tts/sampler.py]: Done. Please review the changes to see if they meet your requirements.

File: sglang_omni/models/higgs_tts/engine_builder.py [hunk 7]
```diff
@@ -83,6 +83,7 @@ def generation_defaults(
                 if self.total_gpu_memory_fraction is not None
                 else 0.85
             ),
+            "sampling_backend": "pytorch",
```
Comment (Ratish1) [inline sglang_omni/models/higgs_tts/engine_builder.py]: NPU already resolves the SGLang sampling backend to PyTorch because FlashInfer is NVIDIA-only. remove this override so the NPU path stays unchanged and CUDA retains its existing auto-selected backend.

File: sglang_omni/models/higgs_tts/stages.py [hunk 8]
```diff
@@ -535,6 +548,15 @@ def create_vocoder_executor(
     # margin; when overriding strides, re-derive the domain empirically.
     checkpoint_dir = resolve_checkpoint(model_path)
     codec = get_or_load_codec(checkpoint_dir, device, dtype)
+    if compile_decode:
+        if current_platform.is_npu():
+            logger.warning(
+                "compile_decode=True was requested but the current platform (%s) "
+                "has no torch.compile backend; falling back to the eager "
```
Comment (Ratish1) [inline sglang_omni/models/higgs_tts/stages.py]: nit: torch-npu has a compile backend; this codec dynamic-shape graph is the unsupported part. should we change this warning to mention that?

File: sglang_omni/models/higgs_tts/stages.py [hunk 9]
```diff
@@ -379,7 +379,7 @@ def _preprocess(payload: StagePayload) -> StagePayload:
 def create_audio_encoder_executor(
     model_path: str,
     *,
-    device: str = "cuda:0",
+    device: str | None = None,
```
Comment (Ratish1) [inline sglang_omni/models/higgs_tts/stages.py]: nit: this stage now runs on the selected accelerator. change `GPU stage` in the docstring and the remaining `same-GPU processes` wording in `config.py` to device-neutral terms.

**Review by Ratish1 — COMMENTED** [review]
Can we fix the lint as well @chx96642264 , thanks

File: sglang_omni/models/higgs_tts/engine_builder.py (hunk 7 above)
Comment (chx96642264) [reply sglang_omni/models/higgs_tts/engine_builder.py]: Done. I have removed the override.

File: sglang_omni/models/higgs_tts/stages.py (hunk 8 above)
Comment (chx96642264) [reply sglang_omni/models/higgs_tts/stages.py]: Done. You're right, this was something I overlooked. It has been fixed.

File: sglang_omni/models/higgs_tts/__init__.py [hunk 10]
```diff
@@ -13,18 +13,23 @@
...
+from sglang_omni.platforms import current_platform
 
 from . import config
 from .hf_config import HiggsMultimodalQwen3Config
 
 AutoConfig.register("higgs_multimodal_qwen3", HiggsMultimodalQwen3Config)
 
+# NPU's compile backend cannot compile this codec model 
+# (crashes on dynamic shapes, verified on Atlas 910B).
+_supports_torch_compile = not current_platform.is_npu()
```
Comment (Ratish1) [inline sglang_omni/models/higgs_tts/__init__.py]: capability lookup is keyed only by model architecture, and the shared capability tests require Higgs to remain `supports_torch_compile=True`. on NPU this import-time value becomes false and contradicts that contract. keep the architecture capability true and leave the NPU-specific disablement at the two actual compile call sites.

Comment (chx96642264) [reply sglang_omni/models/higgs_tts/__init__.py]: Done. I tested and verified this, and your point is indeed valid. This was an addition I made by mistake. 
The NPU-specific disablement stays at the two actual compile call sites in stages.py (the acoustic-encoder gate and the vocoder compile_decode fallback).

**Comment by Ratish1** [comment]
The unit test is failing, could we fix it @chx96642264 , thanks
